listings/ch00-introduction/testing/src/lib.cairo

```diff
@@ -0,0 +1,134 @@
+    use starknet::{get_caller_address, ContractAddress};
+
+    #[storage]
+    struct Storage {
+        value: u32,
+        owner: ContractAddress
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, initial_value: u32) {
+        self.value.write(initial_value);
+        self.owner.write(get_caller_address());
+    }
+
+    #[external(v0)]
+    impl SimpleContract of super::ISimpleContract<ContractState> {
+        fn get_value(self: @ContractState) -> u32 {
+            self.value.read()
+        }
+
+        fn get_owner(self: @ContractState) -> ContractAddress {
+            self.owner.read()
+        }
+
+        fn set_value(ref self: ContractState, value: u32) {
+            assert(self.owner.read() == get_caller_address(), 'Not owner');
+            self.value.write(value);
+        }
+    }
+}
+// ANCHOR_END: contract
+
+// ANCHOR: test
+#[cfg(test)]
+mod simple_contract_tests {
+    // Import the interface and dispatcher to be able to interact with the contract.
+    use super::{
+        ISimpleContract, SimpleContract, ISimpleContractDispatcher, ISimpleContractDispatcherTrait
+    };
+
+    // Import the deploy syscall to be able to deploy the contract.
+    use starknet::class_hash::Felt252TryIntoClassHash;
+    use starknet::{
+        deploy_syscall, ContractAddress, get_caller_address, get_contract_address,
+        contract_address_const
+    };
+
+    // Use debug print trait to be able to print result if needed.
+    use debug::PrintTrait;
+
+    // Use starknet test utils to fake the transaction context.
+    use starknet::testing::{set_caller_address, set_contract_address};
+
+    use serde::Serde;
+    use option::OptionTrait;
+    use array::ArrayTrait;
+    use traits::{Into, TryInto};
+    use result::ResultTrait;
+
+    // Deploy the contract and return its dispatcher.
+    fn deploy(initial_value: u32) -> ISimpleContractDispatcher {
+        // Set up constructor arguments.
+        let mut calldata = ArrayTrait::new();
+        initial_value.serialize(ref calldata);
+
+        // Declare and deploy
+        let (contract_address, _) = deploy_syscall(
+            SimpleContract::TEST_CLASS_HASH.try_into().unwrap(), 0, calldata.span(), false
+        )
+            .unwrap();
+
+        // Return the dispatcher.
+        // The dispatcher allows to interact with the contract based on its interface.
+        ISimpleContractDispatcher { contract_address }
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_deploy() {
+        let initial_value: u32 = 10;
+        let contract = deploy(initial_value);
+
+        assert(contract.get_value() == initial_value, 'wrong initial value');
+        assert(contract.get_owner() == get_contract_address(), 'wrong owner');
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_set_as_owner() {
+        // Fake the caller address to address 1
+        let owner = contract_address_const::<1>();
+        set_caller_address(owner);
+
```

Comment:
does it work? shouldn't be `set_contract_address`?

Reply:
It works that way and not with `set_contract_ address`...
Actually, I'm not 100% sure if I understood the difference between `set_caller_address` and `set_contract_address` in this context.

Reply:
Normally the test is considered as a contract, so to mock it's address you use `set_contract_address`. That way when calling other contracts, `get_caller_address` reflects your test address

Reply:
That's what I understood as well. Well we can let it like this for now.